File: sdks/flutter/lib/src/queue.dart
import 'dart:async';

import 'identity.dart';
import 'types.dart';

/// Same cadence as the web snippet.
const int flushSize = 10;
const int flushIntervalMs = 5000;

/// /v1/events caps a batch at 50 events — split larger buffers.
const int _maxBatch = 50;

/// Buffers events and flushes on size / interval / app-background. The ingest
/// response may carry a pending Live Assist — surfaced via the response handler.
class EventQueue {
  final String _key;
  final String _eventsUrl;
  final Identity _identity;
  final DeviceInfo _device;
  final Transport _transport;
  final Clock _now;

  final List<EventInput> _buffer = [];
  Timer? _timer;
  void Function(Object? data)? _onResponse;
  Future<void> _inflight = Future<void>.value();

  EventQueue(
    this._key,
    this._eventsUrl,
    this._identity,
    this._device,
    this._transport,
    this._now,
  );

  void setResponseHandler(void Function(Object? data) fn) {
    _onResponse = fn;
  }

  void enqueue(EventInput event) {
    _identity.touchSession();
    _buffer.add(event);
    if (_buffer.length >= flushSize) {
      flush();
    } else if (_timer == null) {
      _timer = Timer(
        const Duration(milliseconds: flushIntervalMs),
        () => flush(),
      );
    }
  }

  /// Send everything buffered. Serialized so batches arrive in order.
  Future<void> flush() {
    if (_timer != null) {
      _timer!.cancel();
      _timer = null;
    }
    if (_buffer.isEmpty) return _inflight;
    final events = List<EventInput>.from(_buffer);
    _buffer.clear();
    _inflight = _inflight.then((_) async {
      for (var i = 0; i < events.length; i += _maxBatch) {
        final end = (i + _maxBatch < events.length) ? i + _maxBatch : events.length;
        final batch = Batch(
          key: _key,
          anonId: _identity.getAnonId(),
          userId: _identity.getUserId(),
          sessionId: _identity.currentSession(),
          sentAt: _now(),
          device: _device,
          events: events.sublist(i, end),
        );
        try {
          final data = await _transport.post(_eventsUrl, batch.toJson());
          if (data != null) _onResponse?.call(data);
        } catch (_) {
          // fail-silent: behavioral telemetry must never break the host app
        }
      }
    });
    return _inflight;
  }

  /// Test helper.
  int get size => _buffer.length;
}

File: sdks/android/src/main/kotlin/app/tracki/sdk/Queue.kt
package app.tracki.sdk

import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

/** Same cadence as the web snippet. */
const val FLUSH_SIZE = 10
const val FLUSH_INTERVAL_MS = 5000L

/** /v1/events caps a batch at 50 events — split larger buffers. */
private const val MAX_BATCH = 50

/**
 * Buffers events and flushes on size / interval / app-background. The ingest
 * response may carry a pending Live Assist — surfaced via the response handler.
 */
class EventQueue(
    private val key: String,
    private val eventsUrl: String,
    private val identity: Identity,
    private val device: DeviceInfo,
    private val transport: Transport,
    private val now: Clock,
    private val scope: CoroutineScope,
) {
    private val buffer = ArrayList<EventInput>()
    private var timer: Job? = null
    private var onResponse: ((Any?) -> Unit)? = null
    /** Serializes flushes so batches arrive in order (mirrors the TS inflight chain). */
    private var inflight: Job = scope.launch { }

    fun setResponseHandler(fn: (Any?) -> Unit) {
        onResponse = fn
    }

    fun enqueue(event: EventInput) {
        identity.touchSession()
        val count: Int
        synchronized(buffer) {
            buffer.add(event)
            count = buffer.size
        }
        if (count >= FLUSH_SIZE) {
            flush()
        } else if (timer == null) {
            timer = scope.launch {
                delay(FLUSH_INTERVAL_MS)
                flush()
            }
        }
    }

    /** Send everything buffered. Serialized so batches arrive in order. */
    fun flush(): Job {
        timer?.cancel()
        timer = null
        val events: List<EventInput>
        synchronized(buffer) {
            if (buffer.isEmpty()) return inflight
            events = ArrayList(buffer)
            buffer.clear()
        }
        val prev = inflight
        val next = scope.launch {
            prev.join()
            var i = 0
            while (i < events.size) {
                val end = minOf(i + MAX_BATCH, events.size)
                val batch = Batch(
                    key = key,
                    anonId = identity.getAnonId(),
                    userId = identity.getUserId(),
                    sessionId = identity.currentSession(),
                    sentAt = now(),
                    device = device,
                    events = events.subList(i, end),
                )
                val data = runCatching { transport.post(eventsUrl, batch.toJson()) }.getOrNull()
                // fail-silent: behavioral telemetry must never break the host app.
                if (data != null) onResponse?.invoke(data)
                i += MAX_BATCH
            }
        }
        inflight = next
        return next
    }

    /** Await-friendly flush mirroring the reference's `flush(): Promise<void>`. */
    suspend fun flushAndJoin() {
        flush().join()
    }

    /** Test helper. */
    val size: Int
        get() = synchronized(buffer) { buffer.size }
}
